streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -492,22 +492,18 @@ private void handleStreamsUncaughtException(final Throwable throwable,
                 closeToError();
                 break;
             case SHUTDOWN_APPLICATION:
+                if (getNumLiveStreamThreads() <= 1) {
```

Comment:
This PR LGTM but can you add a quick test for the case of a global-only topology with injected error + SHUTDOWN_APPLICATION handler? Just want to make sure nothing weird happens when you try to start up a StreamThread with no topology

Reply:
I wasn't sure about how to inject an error into an only global topology. But I figured what we really needed to test was if we can add a thread to a global-only topology. So I did that WDYT?

It will hit an illegal state exception as "Consumer is not subscribed to any topics or assigned any partitions" but that is after it will have resized the cache and after the shutdown rebalance has been triggered

Reply:
You can use `StreamsBuilder#addGlobalStore` to inject an error, this method has a required parameter for a Processor which you're supposed to use to insert the data into the store (tbh no idea why you have to implement this yourself, it's caused a LOT of problems since it's unclear what you can/can't do with this -- but that's a whole other can of worms). 

Adding a thread is probably a fine workaround for now, it clearly answered the question, but imo we should just go ahead and use the `StreamsBuilder#addGlobalStore` + injected error to decouple the test from the way we currently happen to handle this. Then we won't have to rewrite it if we ever change something.

Regarding the IllegalSTateException: wouldn't the thread hit this `IllegalStateException` before it got the chance to send the shutdown signal in a rebalance? IIRC the `subscribe` that's throwing is at the very beginning of the StreamThread loop. Also, I take it your test is passing but I wonder why we don't get stuck in an endless cycle? The new thread hits an unexpected exception and invokes the handler, SHUTDOWN_APPLICATION notices that this is the last StreamThread so it starts a new one, and so on.

Reply:
Well we know that the `SHUTDOWN_APPLICATION` will fail because the other client will not have a thread to receive the signal, unless you mean applications with different topologies?

It doesn't appear to endlessly loop the illegal state exception because both threads are still alive at that point but  it does not work as intended. I think that maybe this won't solve the global thread only case. We should only add a stream thread when there is one thread but not if there is zero. At least this ill take care of the eos problem and we can handle the global case as before

Reply:
Also the `IllegalStateException` is coming from the runLoop not the `subcribe`

Reply:
Ah, I guess it's from trying to poll without any subscribed topics. Since that's not the case this alleviates my concern about the endless cycle too (In hindsight, yeah, no way `subscribe(<empty topics>)` would throw since that's supposed to be semantically identical to calling `unsubscribe()`)